ccmlib/node.py

```diff
@@ -558,21 +558,35 @@ def watch_log_for(self, exprs, from_mark=None, timeout=600, process=None, verbos
                             matchings.append((line, m))
                             tofind.remove(e)
                             if len(tofind) == 0:
+                                common.debug("return")
                                 return matchings[0] if isinstance(exprs, string_types) else matchings
                 else:
                     # yep, it's ugly
                     time.sleep(1)
                     if start + timeout < time.time():
                         raise TimeoutError(time.strftime("%d %b %Y %H:%M:%S", time.gmtime()) + " [" + self.name + "] Missing: " + str([e.pattern for e in tofind]) + ":\n" + reads[:50] + ".....\nSee {} for remainder".format(filename))
 
-                if process:
-                    if common.is_win():
-                        if not self.is_running():
-                            return None
-                    else:
-                        process.poll()
-                        if process.returncode == 0:
-                            return None
+                    # update status because meanwhile node could have died....
+                    # if it was running but now it is not we should return immediately rather than after timeout
+                    if self.pid and not self._is_pid_running():
+                        msg = "node {n} with pid {pid} stopped during watching log for {m}".\
+                            format(n=self.name, pid=self.pid, m=exprs)
+                        common.warning(msg)
+                        raise NodeError(msg)
+
+                # process is in this case process that started C*
```

Comment:
This is another tricky part....

Possibly the intention of this part of the code was to actually finish immediately if child process terminated... 
`return None` as not supper effective as results are not checked.

There is one interesting thing:
when checking `None` it can fail positive tests: `test3` or `test2` which is kind of unexpected.